src/FastConv/FastConv.cpp

```diff
@@ -1,32 +1,192 @@
+    CFft::createInstance(fft);
+    fft->initInstance(blockLength*2, 1, CFft::kWindowHann, CFft::kNoWindow);
+    for (int i = 0; i < numBlocks; i++) {
+        int thisBlockLength = std::min(blockLength, length - i * blockLength);
+        float *impulseResponseBlock = &impulseResponseBlocks[i * blockLength*2];
+        std::copy_n(&impulseResponse[i * blockLength], thisBlockLength, impulseResponseBlock);
+        fft->doFft(impulseResponseBlock, impulseResponseBlock);
+    }
+    // We scale by the FFT length once, here, to avoid doing in repeatedly in `addMultiplySpectra`.
+    CVectorFloat::mulC_I(impulseResponseBlocks.data(), blockLength * 2, impulseResponseBlocks.size());
+}
+
+FreqConvolution::~FreqConvolution() {
+    CFft::destroyInstance(fft);
+}
+
+// Time-domain implementation of circular convolution, for reference/testing:
+// void circularConvolve(float *output, const float *a, const float *b, int length) {
+//     for (int i = 0; i < length; i++) {
+//         float acc = 0;
+//         for (int j = 0; j < length; j++) {
+//             acc += a[j] * b[(i - j + length) % length];
+//         }
+//         output[i] = acc;
+//     }
+// }
+
+// Previous implementation of `multiplySpectra`, included for reference.
+// This has the advantage of not needing information about the internal layout used by CFft;
+// however, the repeated splitRealImag/mergeRealImag calls add considerable overhead.
+
+// void FreqConvolution::multiplySpectra(float *spectrumC, const float *spectrumA, const float *spectrumB) {
+//     std::vector<float> realA(blockLength+1), imagA(blockLength+1);
+//     fft->splitRealImag(realA.data(), imagA.data(), spectrumA);
+
+//     std::vector<float> realB(blockLength+1), imagB(blockLength+1);
+//     fft->splitRealImag(realB.data(), imagB.data(), spectrumB);
+
+//     float scale = blockLength * 2;
+//     for (int i = 0; i < blockLength+1; i++) {
+//         float tmp = realA[i];
+//         realA[i] = (tmp * realB[i] - imagA[i] * imagB[i]) * scale;
+//         imagA[i] = (tmp * imagB[i] + imagA[i] * realB[i]) * scale;
+//     }
+//     fft->mergeRealImag(spectrumC, realA.data(), imagA.data());
+// }
+
+// Helper function: multiply two complex spectra, add result to output.
+static inline void addMultiplySpectra(float *output, const float *spectrumA, const float *spectrumB, int blockLength) {
+    // Spectrum layout: re(0),re(1),re(2),...,re(size/2),im(size/2-1),...,im(1)
+    output[0] += spectrumA[0] * spectrumB[0];
+    output[blockLength] += spectrumA[blockLength] * spectrumB[blockLength];
+    for (int i = 1; i < blockLength; i++) {
+        float realA = spectrumA[i];
+        float imagA = spectrumA[blockLength*2-i];
+        float realB = spectrumB[i];
+        float imagB = spectrumB[blockLength*2-i];
+        output[i] += realA * realB - imagA * imagB; // Real part
+        output[blockLength*2-i] += realA * imagB + imagA * realB; // Imaginary part
+    }
```

Comment:
It seems retrieving the FFT results directly is more efficient! (I suppose)

Reply:
Yeah, I did some profiling with valgrind and found that a surprisingly large chunk of time was spent just splitting and merging the real/imag parts.

(I'd prefer to split/merge to avoid needing to know implementation details of `CFft`, but this is a case where there's a tension between abstraction and performance. Ideally, `CFft` or a related class would provide some methods for doing math on its complex array representation.)